src/source.rs

```diff
@@ -8,104 +14,282 @@ use bevy::{
-            _ => {}
-        }
+impl Decodable for SteamAudio {
+    type Decoder = SteamDecoder;
+
+    type DecoderItem = <SteamDecoder as Iterator>::Item;
+
+    fn decoder(&self) -> Self::Decoder {
+        SteamDecoder::new(self.direction.clone(), self.path.clone())
     }
 }
 
-pub struct Listener(pub Entity);
+// Todo implement default
+#[derive(Resource)]
+pub struct SpatialAudioSettings {
+    pub audio_settings: AudioSettings,
+    pub context_settings: ContextSettings,
+    pub hrtf_settings: HRTFSettings,
+    pub simulation_settings: SimulationSettings,
+    pub context: Context,
+    pub hrtf: HRTF,
+    pub simulator: Simulator,
+}
+
+pub struct SpatialAudioPlugin;
+
+impl Plugin for SpatialAudioPlugin {
+    fn build(&self, app: &mut App) {
+        let audio_settings = AudioSettings::default();
+        let context_settings = ContextSettings::default();
+        let hrtf_settings = HRTFSettings::default();
+        let simulation_settings = SimulationSettings::from_audio_settings(&audio_settings);
+
+        let context = Context::new(&context_settings).expect("could not build steam audio context");
+        let hrtf = HRTF::new(&context, &audio_settings, &hrtf_settings)
+            .expect("could not build steam audio hrtf");
+        let simulator = Simulator::new(&context, &simulation_settings)
+            .expect("could not build steam audio simulation");
+
+        app.insert_resource(SpatialAudioSettings {
+            audio_settings,
+            context_settings,
+            hrtf_settings,
+            simulation_settings,
+            context,
+            hrtf,
+            simulator,
+        });
+    }
+}
+
+// pub fn context_update(mut commands: Commands, settings: Res<ContextSettings>) {
+//     if settings.is_changed() {
+//         match Context::new(&*settings) {
+//             Ok(context) => {
+//                 commands.insert_resource(context);
+//             }
+//             _ => {}
+//         }
+//     }
+// }
+
+// pub fn hrtf_update(
+//     mut commands: Commands,
+//     context: Res<Context>,
+//     audio_settings: Res<AudioSettings>,
+//     hrtf_settings: Res<HRTFSettings>,
+// ) {
+//     if context.is_changed() || audio_settings.is_changed() || hrtf_settings.is_changed() {
+//         match HRTF::new(&context, &audio_settings, &hrtf_settings) {
+//             Ok(hrtf) => {
+//                 commands.insert_resource(hrtf);
+//             }
+//             _ => {}
+//         };
+//     }
+// }
+
+// pub fn simulation_update(
+//     mut commands: Commands,
+//     context: Res<Context>,
+//     simulation_settings: Res<SimulationSettings>,
+// ) {
+//     if context.is_changed() || simulation_settings.is_changed() {
+//         match Simulator::new(&*context, &simulation_settings) {
+//             Ok(simulator) => {
+//                 commands.insert_resource(simulator);
+//             }
+//             _ => {}
+//         }
+//     }
+// }
+
+// Todo: Need to get rid of this...
```

Comment:
We probably just need to make steam-audio-rs match bevy's glam version here

Reply:
Actually maybe we shouldn't even use `glam` in steam-audio-rs in the first place and just use `[f32; 3]`? We don't really do any math operations with them ourselves since it's just a wrapper. That way we could just directly use `bevy::math::Vec3` here.

Reply:
> Actually maybe we shouldn't even use `glam` in steam-audio-rs in the first place and just use `[f32; 3]`? We don't really do any math operations with them ourselves since it's just a wrapper. That way we could just directly use `bevy::math::Vec3` here.

Yeah that sounds reasonable. I can take a stab at this in the weekend if you don't get to it before me.

Reply:
@Aceeri looks a lot better now IMO